usrsctplib/netinet/sctp_callout.c

```diff
@@ -75,72 +76,201 @@ uint32_t sctp_get_tick_count(void) {
 
 /*
  * SCTP_TIMERQ_LOCK protects:
- * - SCTP_BASE_INFO(callqueue)
- * - sctp_os_timer_next: next timer to check
+ * - SCTP_BASE_INFO(timers_queue)
  * - sctp_os_timer_current: current callout callback in progress
  * - sctp_os_timer_current_tid: current callout thread id in progress
- * - sctp_os_timer_waiting: some thread is waiting for callout to complete
- * - sctp_os_timer_wait_ctr: incremented every time a thread wants to wait
- *                           for a callout to complete.
+ * - sctp_os_timer_current_changed: conditional variable signaled when
+ *                                  current callout pointer is changed
  */
-static sctp_os_timer_t *sctp_os_timer_next = NULL;
 static sctp_os_timer_t *sctp_os_timer_current = NULL;
-static int sctp_os_timer_waiting = 0;
-static int sctp_os_timer_wait_ctr = 0;
 static userland_thread_id_t sctp_os_timer_current_tid;
+static userland_cond_t sctp_os_timer_current_changed;
 
-/*
- * SCTP_TIMERWAIT_LOCK (sctp_os_timerwait_mtx) protects:
- * - sctp_os_timer_wait_cond: waiting for callout to complete
- * - sctp_os_timer_done_ctr: value of "wait_ctr" after triggering "waiting"
- */
-userland_mutex_t sctp_os_timerwait_mtx;
-static userland_cond_t sctp_os_timer_wait_cond;
-static int sctp_os_timer_done_ctr = 0;
+#if defined(__Userspace__)
+
+static void
+sctp_userland_cond_wait(userland_cond_t* cond,
+                        userland_mutex_t* mtx) {
+	// callee must be robust to spurious wakeups, both because
+	// wakeup could happen due to native API behavior and because
+	// there is indendend timeout for wait.
+#if defined(__Userspace_os_Windows)
+	const DWORD timeoutMillis = 20 * 1000;
+	const BOOL waited = SleepConditionVariableCS(cond, mtx, timeoutMillis);
+	if (!waited) {
+		SCTP_PRINTF("WARN; SleepConditionVariableCS did not return within %ul millis\n", timeoutMillis);
```

Comment:
Note for reviewers:
I think this `printf` and non-infinite wait timeout could be useful for future debugging, whenever someone misuse blocking `sctp_os_timer_stop` when some mutex necessary to execute timer's callback is held (deadlock condition).